### PR TITLE
Enable `--resource-type` and `--exclude-resource-type` CLI flags and environment variables for `dbt test`

### LINKED ISSUE #10656: [Feature] Exclude unit tests from running when using "dbt test" command
### Is this your first time submitting a feature request?

- [X] I have read the [expectations for open source contributors](https://docs.getdbt.com/docs/contributing/oss-expectations)
- [X] I have searched the existing issues, and I could not find an existing issue for this feature
- [X] I am requesting a straightforward extension of existing dbt functionality, rather than a Big Idea better suited to a discussion

### Describe the feature

This is somewhat of an extension of a previous issue: https://github.com/dbt-labs/dbt-core/issues/9237

Currently, including "unit_test" in the DBT_EXCLUDE_RESOURCE_TYPES env variable will prevent units from running when using the dbt build, list and clone commands. **No other commands are affected by this env variable.**

However, I want to extend this to include the dbt test command.

The use case is something that we're hitting currently, where we run dbt on our airflow instances; We do not want to run unit tests on our Prod airflow instances. This is recommended in the dbt docs here: https://docs.getdbt.com/docs/build/unit-tests#when-to-run-unit-tests

Some of our legacy dags / code do not utilise dbt build. Rather, they use dbt run and dbt test as separate commands.
I know that there are ways to limit what is run using specific cli commands, but I think an env variable would be much cleaner, and would prevent us from having to change our code in (potentially) multiple places.

Looking at the ticket I posted above, there seems to be an assumption that no one would want to prevent unit tests on a dbt project level, which I think is true. However, as it currently stands, preventing unit_tests on an environment level is not yet 100% achievable across all functions of dbt.

It would be ideal if we could prevent all our dbt commands from running unit tests with a single env variable, that we could set on whatever airflow instances we need.

### Describe alternatives you've considered

We could change all our dags to use dbt build instead of dbt run and dbt test separately.

I believe we could also add an argument to our dbt command to something like this: dbt test --select "test_type:data".

With both these cases, we'd have to change the way we use dbt. Ideally, a global exclusion of a resource_type should be applicable to all functions of dbt.

### Who will this benefit?

Anyone who wants to prevent unit tests from running on a per environment basis.

### Are you interested in contributing this feature?

I am just about to create a PR for this.

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by dbeatty10**
Resolves #10656

Alternative to #10657

### Problem

Currently, the [resource type](https://docs.getdbt.com/reference/global-configs/resource-type) CLI flags and environment variables apply to the `dbt build`, `dbt list` and `dbt clone` subcommands **ONLY** -- they do not extend to `dbt test`.

We want to extend these to apply to the `dbt test` subcommand also.

### Solution

Simplify by standardizing upon the `ResourceTypeSelector` which already provides the desired `resource_types` parameter.

This was accomplished by:
1. adding the `resource_type` and `exclude_resource_type` click args to the `dbt test` subcommand
1. adding a `TEST_NODE_TYPES` list of valid test types (data tests and unit tests)
1. 👉 adding `resource_types` property to the `TestTask` (which uses `TEST_NODE_TYPES`) 👈 **this was the crux!**

### Backstory

Branched from https://github.com/dbt-labs/dbt-core/pull/10657 to try an alternative approach that resolves the issue via a refactor.

Did these PRs leading up to this one:
- https://github.com/dbt-labs/dbt-core/pull/10730
- https://github.com/dbt-labs/dbt-core/pull/10718

### Code examples

dbt `test` now enables passing values like `unit_tests`, `tests`, etc. in the `--resource-type` / `--exclude-resource-type` CLI flags (or in the  environment variables `DBT_RESOURCE_TYPES` / `DBT_EXCLUDE_RESOURCE_TYPES`).

#### Example 1 - CLI flags

```shell
dbt test --resource-type unit_test
dbt test --exclude-resource-type unit_test
```

#### Example 2 - environment variables

```shell
DBT_RESOURCE_TYPES=unit_test dbt test
DBT_EXCLUDE_RESOURCE_TYPES=unit_test dbt test
```

### Documentation to update

At minimum (there may be other relevant pages):
- https://docs.getdbt.com/reference/global-configs/resource-type

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me.
- [x] I have run this code in development, and it appears to resolve the stated issue.
- [x] This PR includes tests, or tests are not required or relevant for this PR.
- [x] This PR has no interface changes (e.g., macros, CLI, logs, JSON artifacts, config files, adapter interface, etc.) or this PR has already received feedback and approval from Product or DX.
- [x] This PR includes [type annotations](https://docs.python.org/3/library/typing.html) for new and modified functions.

### CODE REVIEW

File: .changes/unreleased/Features-20240903-132428.yaml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+kind: Features
+body: Adding logic to TestSelector to remove unit tests if they are in excluded_resource_types
+time: 2024-09-03T13:24:28.592837+01:00
+custom:
+    Author: TowardOliver
```
Comment (dbeatty10) [inline .changes/unreleased/Features-20240903-132428.yaml:5]:
```suggestion
    Author: TowardOliver dbeatty10
```

File: .changes/unreleased/Features-20240903-132428.yaml [hunk 2]
```diff
@@ -0,0 +1,6 @@
+kind: Features
+body: Adding logic to TestSelector to remove unit tests if they are in excluded_resource_types
```
Comment (dbeatty10) [inline .changes/unreleased/Features-20240903-132428.yaml:2]:
```suggestion
body: Enable `--resource-type` and `--exclude-resource-type` CLI flags and environment variables for `dbt test`
```

File: core/dbt/task/test.py [hunk 3]
```diff
@@ -387,14 +397,24 @@ class TestTask(RunTask):
...
+    @property
+    def resource_types(self) -> List[NodeType]:
+        resource_types: Collection[NodeType] = resource_types_from_args(
+            self.args, set(TEST_NODE_TYPES), set(TEST_NODE_TYPES)
+        )
+
+        # filter out any non-test node types
+        resource_types = [rt for rt in resource_types if rt in TEST_NODE_TYPES]
+        return list(resource_types)
+
```
Comment (dbeatty10) [inline core/dbt/task/test.py:409]: This is basically the same thing as [here](https://github.com/dbt-labs/dbt-core/blob/3e437a67344469d51d7f2d4a4d0338a45c1de261/core/dbt/task/list.py#L179-L188) and [here](https://github.com/dbt-labs/dbt-core/blob/3e437a67344469d51d7f2d4a4d0338a45c1de261/core/dbt/task/build.py#L85-L94).

File: core/dbt/task/test.py [hunk 4]
```diff
@@ -387,14 +397,24 @@ class TestTask(RunTask):
...
     def get_node_selector(self) -> ResourceTypeSelector:
         if self.manifest is None or self.graph is None:
             raise DbtInternalError("manifest and graph must be set to get perform node selection")
         return ResourceTypeSelector(
             graph=self.graph,
             manifest=self.manifest,
             previous_state=self.previous_state,
-            resource_types=[NodeType.Test, NodeType.Unit],
+            resource_types=self.resource_types,
```
Comment (dbeatty10) [inline core/dbt/task/test.py:417]: Basically same thing as [here](https://github.com/dbt-labs/dbt-core/blob/3e437a67344469d51d7f2d4a4d0338a45c1de261/core/dbt/task/list.py#L214) and [here](https://github.com/dbt-labs/dbt-core/blob/3e437a67344469d51d7f2d4a4d0338a45c1de261/core/dbt/task/build.py#L210).

**Review by gshank — APPROVED** [review]

**Pull request merged**